Add optional Avro namespace to CREATE SINK #8352

pH14 opened the pull request:
<!--

Describe the contents of the PR briefly but completely.

If you write detailed commit messages, it is acceptable to copy/paste them
here, or write "see commit messages for details." If there is only one commit
in the PR, GitHub will have already added its commit message above.

-->

Adds an optional `avro_class_name` field that will be used for generated Avro schemas produced by `CREATE SINK`. This change here is mostly plumbing the input value to the `name` field in our existing Avro schema serde, which already knows how to properly handle name/namespace splitting if the input has dots (e.g. `abc.def.ghi => {name: ghi, namespace: abc.def`)

Testing is more interesting, and this PR adds a new `kafka-verify-schema` action that's used to validate that the generated key/value schemas are exactly what we expect.

### Motivation

Fixes #8352

<!--

Which of the following best describes the motivation behind this PR?

  * This PR fixes a recognized bug: [Link to issue.]

  * This PR adds a known-desirable feature: [Link to issue.]

  * This PR fixes a previously unreported bug.

    [Describe the bug in detail, as if you were filing a bug report.]

  * This PR adds a feature that has not yet been specified.

    [Write a brief specification for the feature, including justification
     for its inclusion in Materialize, as if you were writing the original
     feature specification.]

   * This PR refactors existing code.

    [Describe what was wrong with the existing code, if it is not obvious.]

-->

### Tips for reviewer

Both a Rust and Materialize rookie here, so no amount of nitpicking will go unappreciated 😁 
<!--

Leave some tips for your reviewer, like:

    * The diff is much smaller if viewed with whitespace hidden.
    * [Some function/module/file] deserves extra attention.
    * [Some function/module/file] is pure code movement and only needs a skim.

Delete this section if no tips.

-->

### Checklist

(These are mostly TODO. Will complete these before unmarking the PR as draft)

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR adds a release note for any [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

pH14 commented on src/testdrive/src/action.rs:
```diff
@@ -441,7 +441,10 @@ pub async fn build(cmds: Vec<PosCommand>, state: &State) -> Result<Vec<PosAction
                         Box::new(kafka::build_verify(builtin, context.clone()).map_err(wrap_err)?)
-                    }
+                    },
+                    "kafka-verify-schema" => {
```
tentatively put this under `kafka` rather than a more generic schema check action... the schemas get unique names each test run, and they're resolved by querying the `mz_kafka_sinks` table like the existing `kafka-verify` does

umanwizard commented on src/testdrive/src/action/kafka/verify.rs:
```diff
@@ -273,3 +274,98 @@ impl Action for VerifyAction {
+    };
+    let sink = cmd.args.string("sink")?;
+
+    let (key, value) : (Option<String>, String ) = match &cmd.input[..] {
+        [value] => (None, value.clone()),
+        [key, value] => (Some(key.clone()), value.clone()),
+        _ => return Err(String::from("unable to read key/value schema inputs"))
+    };
+
```
There are a few extra spaces here. You should run `cargo fmt`, otherwise your change will be rejected by CI.

umanwizard commented on src/testdrive/src/action/kafka/verify.rs:
```diff
@@ -273,3 +274,98 @@ impl Action for VerifyAction {
+                    }
+                }
+            },
+            _ => unimplemented!()
```
Could you add an error message here, like `unimplemented!("kafka-verify-schema is only supported for Avro sinks")`

umanwizard reviewed (commented):
This looks mostly good to me, just some nits.

umanwizard commented:
Somehow I accidentally deleted your comment. I think it's fine to have the verification command be Kafka-specific for now.

umanwizard commented on src/interchange/src/avro/encode.rs:
```diff
@@ -159,10 +160,10 @@ impl AvroSchemaGenerator {
         let key_info = key_desc.map(|key_desc| {
             let columns = column_names_and_types(key_desc);
-            let row_schema = build_row_schema_json(&columns, "row");
+            let row_schema = build_row_schema_json(&columns, &class_name.unwrap_or("row".to_string()));
```
Also, even though it will no longer be relevant after you make the change I suggested, you should never use `unwrap_or` with a function that does a nontrivial amount of work (like cloning, allocating memory, etc.), which includes `to_string`. The reason is that it is evaluated eagerly, so the work is being done regardless of whether the option is `None` or `Some`. 

I'm pretty sure we have a lint rule that enforces this, actually.

Instead, you can use `unwrap_or_else`, which accepts a function that is only called if it actually needs to be; e.g. instead of `class_name.unwrap_or("row".to_string())` you would write `class_name.unwrap_or_else(|| "row".to_string())`.

pH14 commented on src/sql/src/plan/statement/ddl.rs:
```diff
@@ -1433,6 +1433,12 @@ fn kafka_sink_builder(
     };
     let config_options = kafka_util::extract_config(with_options)?;
 
+    let avro_class_name = match with_options.remove("avro_class_name") {
```
🙇 good catch here

pH14 commented on src/sql/src/plan/statement/ddl.rs:
```diff
@@ -1433,6 +1433,30 @@ fn kafka_sink_builder(
+
+    if key_desc_and_indices.is_some() {
+        match (avro_key_fullname.is_some(), avro_value_fullname.is_some()) {
+            (true, false) | (false, true) =>
```
this whole bit is an exclusive or, wasn't sure what the most idiomatic way to write it would be. I've never found `^` to be super readable, but that might just be from unfamiliarity

umanwizard replied:
I would have written `!(a ^ b)` personally but either is fine IMO, as is `(a && b) || !(a || b)`.

pH14 commented:
* Split the option into `avro_key_fullname` and `avro_value_fullname` so they can/must be independently specified
* Added tests to cover the new scenarios
* Added some release notes + user facing docs updates (cc @morsapaes if you wanted to take a quick look!)

umanwizard commented on doc/user/content/sql/create-sink.md:
```diff
@@ -68,6 +68,8 @@ Field                | Value type | Description
 `retention_ms`       | `long`     | Sets the maximum time Kafka will retain a log.  Accepts values [-1, ...]. `-1` specifics no time limit.  If not set, uses the broker default.
 `retention_bytes`    | `long`     | Sets the maximum size a Kafka partion can grow before removing old logs.  Accepts values [-1, ...]. `-1` specifics no size limit.  If not set, uses the broker default.
+`avro_key_fullname`  | `text`     | Sets the Avro fullname on the generated key schema, if a `KEY` is specified. When used, a value must be specified for `avro_value_fullname`. The default fullname is `row`.
+`avro_value_fullname`| `text      | Sets the Avro fullname on the generated value schema. The default fullname is `envelope`.
```
```suggestion
`avro_value_fullname`| `text`      | Sets the Avro fullname on the generated value schema. The default fullname is `envelope`.
```

umanwizard replied:
Also we should probably mention that value fullname requires key fullname to be specified (if a key exists).

umanwizard reviewed (approved):
LGTM!

umanwizard commented on src/testdrive/src/action/kafka/verify.rs:
```diff
@@ -281,3 +278,100 @@ impl Action for VerifyAction {
+                    }
+                }
+            }
+            _ => unimplemented!("kafka-verify-schema is only supported for Avro sinks"),
```
Maybe it’s better to return an error here rather than crashing Testdrive. @benesch any opinions?

benesch replied:
Yeah, an error is def better!

morsapaes reviewed (commented):
Thanks for the ping, @pH14! Just pushed some changes to the `CREATE SINK` page too, so conflicts incoming for one of us! ⚔️
